packages/tldraw/src/state/data/filesystem.ts

```diff
@@ -1,14 +1,15 @@
+import { fileOpen, fileSave } from 'browser-fs-access'
+import type { FileSystemHandle } from 'browser-fs-access'
 import { get as getFromIdb, set as setToIdb } from 'idb-keyval'
 import { IMAGE_EXTENSIONS, VIDEO_EXTENSIONS } from '~constants'
 import type { TDDocument, TDFile } from '~types'
-import type { FileSystemHandle } from './browser-fs-access'
 
 const options = { mode: 'readwrite' as const }
 
-const checkPermissions = async (handle: FileSystemHandle) => {
+const checkPermissions = async (handle: FileSystemFileHandle) => {
   return (
-    (await handle.queryPermission(options)) === 'granted' ||
-    (await handle.requestPermission(options)) === 'granted'
+    (await (handle as unknown as FileSystemHandle).queryPermission(options)) === 'granted' ||
+    (await (handle as unknown as FileSystemHandle).requestPermission(options)) === 'granted'
```

Comment:
Part of the problem here is that you are mixing global types with the ponyfills from the `browser-fs-access`. I think that it would make sense for you to export your own `FileSystemFileHandle` (one that would extend your `FileSystemHandle`) from the `browser-fs-access`. This is required by the ponyfill nature of that package.

Reply:
I don’t actually modify the handles (`FileSystemFileHandle`, `FileSystemDirectoryHandle`, both under `FileSystemHandle`; they’re just exported because earlier TypeScript versions didn’t know them yet), just the `File` objects. See https://github.com/GoogleChromeLabs/browser-fs-access/blob/main/index.d.ts.

Reply:
`FileSystemFileHandle` and `FileSystemDirectoryHandle` are not even exported from `browser-fs-access`. You reference them in your `index.d.ts` but they are grabbed from `lib.dom.d.ts` (from TypeScript libs). However, your intention is to extend them (sort of).

Conceptually they both extend `FileSystemHandle` and this is a type that you export. Your type is different from the one with the same name in `lib.dom.d.ts` because you add two methods to it (`queryPermission` and `requestPermission`). Both of those should be available for `FileSystemFileHandle` and `FileSystemDirectoryHandle`. 

You could just extend those global types from within your `index.d.ts` and this would start working without any casts here. However, you state that your library is a ponyfill and not a polyfill - so I presume that you don't want to affect a global environment like that. That's why I suggest exporting `FileSystemFileHandle` and `FileSystemDirectoryHandle` from your libraries. If you don't do it then people have to use the global ones for those things and those are, unfortunately, incompatible with your own `FileSystemHandle`.

Reply:
I technically only add stuff the spec has, but which the TypeScript versions don’t yet:

- https://wicg.github.io/file-system-access/#api-filesystemhandle-querypermission
- https://wicg.github.io/file-system-access/#api-filesystemhandle-requestpermission

What’s the recommended way to do this?

Reply:
Yes, I understand why you are doing this. However, as long as they don't add it then u can't count on them being there. And even if they do it eventually (once the feature gets past the experimental stage, I guess) then you might still want to support older versions of TS by providing your own exports.

Alternatively, you could provide people a `.d.ts` file that would extend the global interfaces with what you need. This way users on a new version of TS wouldn't have to do a thing while the ones using older versions would have to include a reference to your augmenting file somehow (for example by including a comment like this in their project: `/// <reference types="browser-fs-access/extend-global-interfaces" />`).

Actually, you could use the last mentioned strategy even today if you want. I just still assume that because your library is acting as a ponyfill you don't want/shouldnt do it. The problem is that, from what I understand, people might get access to those objects in two ways:
1. by using your library - this way those methods are guaranteed to exist
2. by using browser APIs - those methods are not guaranteed to exist

So since those methods are not guaranteed in the second case, you might not want to accept objects obtained this way by your API as your implementation might expect those methods to exist and for them to be callable.

I also understand, that there is always a disconnect in TypeScript between what the typechecker thinks is available and what is actually available at runtime since the browser's runtime and the typechecker are decoupled. Those APIs are still experimental though, so I assume that it's not that uncommon for those not to be available in browsers. Otherwise, there wouldn't be much need for your ponyfill library, right?